### ISSUE TITLE
reportOptionalMemberAccess complaint in list comprehension

### ISSUE THREAD

**Issue by tj-smith47** [microsoft/pyright#9186]
**Describe the bug**
When doing exclusion inside list comprehension, the left side should only run on entities in the list that match the exclusion on the right. If I'm grabbing objects that are dictionaries, then with those I'm getting the value of a key from that dictionary, I shouldn't get a warning that the value on the left could be `None`.

**Code or Screenshots**
```python
commit_signatures = [
    commit.get("signature").get("state", "") for commit in commits if isinstance(commit.get("signature"), dict)
]
```
This throws an error that `commit.get("signature")` could be `None`. The signature field is from a `TypedDict`, defined as having a value of `Optional[dict]`. I understand the source of the error as defined by the `TypedDict` definition, but not why the value on the left isn't seen as being guaranteed to be of type `dict`.

**Comment by erictraut** [comment]
The sample you've provided above refers to the symbol `commits`, which is not defined. Based on your description, I can guess at the missing pieces, but in the future, I'd appreciate it if you'd provide a self-contained code sample to eliminate the need to guess at the missing parts.

The behavior you're seeing is correct, so I don't consider this a bug. Static type checkers cannot assume that a function or method will return the same value every time it is called. The type of the expression `commit.get("signature")` therefore cannot be narrowed to `dict` even though it is included in an `isinstance` conditional.

If you want your code to be type safe, you can assign the value to a temporary value using a walrus expression and use that variable in the outer part of the comprehension.

```python
commit_signatures = [x.get("state", "") for commit in commits if isinstance(x := commit.get("signature"), dict)]
```

**Comment by tj-smith47** [comment]
@erictraut I see, thanks for the followup and recommendation!
